storage/listv2.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storage
+
+import "errors"
+
+// More indicates if the result was truncated. If false
+// then the result []ListItem includes all requested keys.
+// If true then the caller must call List again to get more
+// results by setting `StartAfter` or `EndBefore` appropriately.
+type More bool
+
+// ListOptions are items that are optional for the LIST method
+type ListOptions struct {
+	Prefix       Key
+	StartAfter   Key
+	EndBefore    Key
+	Recursive    bool
+	IncludeValue bool
+	Limit        Limit
+}
+
+// ListV2 lists all keys corresponding to ListOptions
+func ListV2(store KeyValueStore, opts ListOptions) (result Items, more More, err error) {
+	if opts.StartAfter != nil && opts.EndBefore != nil {
+		return nil, false, errors.New("start-after and end-before cannot be combined")
```

Comment:
Why not? Is there a technical limitation for this restriction?

Reply:
Not a technical one, but a semantic one:

```
ListOptions{StartAfter: "", EndBefore: "z", Limit: 3} // last 3
ListOptions{StartAfter: "a", EndBefore: "", Limit: 3} // first 3
ListOptions{StartAfter: "a", EndBefore: "z", Limit: 3} // middle 3?
```

So this either means, there's a need to introduce an additional flag for reverse, or pick one of the orderings for the case when both are present. I think using either behavior, would introduce bugs. But I don't feel too strongly about this.